pin.go

```diff
@@ -0,0 +1,216 @@
+// and an authorization policy that permits TPM2_NV_Increment with a signed authorization policy, signed by the key associated with
+// updateKeyName.
+func createPinNVIndex(tpm *tpm2.TPMContext, handle tpm2.Handle, updateKeyName tpm2.Name, hmacSession tpm2.SessionContext) (*tpm2.NVPublic, tpm2.DigestList, error) {
+	initKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("cannot create signing key for initializing NV index: %w", err)
+	}
+
+	initKeyPublic := createPublicAreaForRSASigningKey(&initKey.PublicKey)
+	initKeyName, err := initKeyPublic.Name()
+	if err != nil {
+		return nil, nil, xerrors.Errorf("cannot compute name of signing key for initializing NV index: %w", err)
+	}
+
+	nameAlg := tpm2.HashAlgorithmSHA256
+
+	// The NV index requires 4 policies:
+	// - A signed policy for initialization, signed with a key that is discarded so that the index cannot be recreated.
+	// - A signed policy for updating the index to revoke old dynamic authorization policies.
+	// - A policy for updating the authorization value (PIN / passphrase).
+	// - A policy to read the counter value without knowing the authorization value, as the value isn't secret.
+	// - A policy to use the count value in a PolicyNV assertion.
+	var authPolicies tpm2.DigestList
+
+	trial, _ := tpm2.ComputeAuthPolicy(nameAlg)
+	trial.PolicyCommandCode(tpm2.CommandNVIncrement)
+	trial.PolicyNvWritten(false)
+	trial.PolicySigned(initKeyName, nil)
+	authPolicies = append(authPolicies, trial.GetDigest())
+
+	trial, _ = tpm2.ComputeAuthPolicy(nameAlg)
+	trial.PolicyCommandCode(tpm2.CommandNVIncrement)
+	trial.PolicyNvWritten(true)
+	trial.PolicySigned(updateKeyName, nil)
+	authPolicies = append(authPolicies, trial.GetDigest())
+
+	trial, _ = tpm2.ComputeAuthPolicy(nameAlg)
+	trial.PolicyCommandCode(tpm2.CommandNVChangeAuth)
+	trial.PolicyAuthValue()
+	authPolicies = append(authPolicies, trial.GetDigest())
+
+	trial, _ = tpm2.ComputeAuthPolicy(nameAlg)
+	trial.PolicyCommandCode(tpm2.CommandNVRead)
+	authPolicies = append(authPolicies, trial.GetDigest())
+
+	trial, _ = tpm2.ComputeAuthPolicy(nameAlg)
+	trial.PolicyCommandCode(tpm2.CommandPolicyNV)
+	authPolicies = append(authPolicies, trial.GetDigest())
+
+	trial, _ = tpm2.ComputeAuthPolicy(nameAlg)
+	trial.PolicyOR(authPolicies)
+
+	// Define the NV index
+	public := &tpm2.NVPublic{
+		Index:      handle,
+		NameAlg:    nameAlg,
+		Attrs:      tpm2.NVTypeCounter.WithAttrs(tpm2.AttrNVPolicyWrite | tpm2.AttrNVAuthRead | tpm2.AttrNVPolicyRead),
+		AuthPolicy: trial.GetDigest(),
+		Size:       8}
+
+	index, err := tpm.NVDefineSpace(tpm.OwnerHandleContext(), nil, public, hmacSession)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("cannot define NV space: %w", err)
+	}
+
+	// NVDefineSpace was integrity protected, so we know that we have an index with the expected public area at the handle we specified
+	// at this point.
+
+	succeeded := false
+	defer func() {
+		if succeeded {
```

Comment:
(nitpick) I guess you could as well check for "if err == nil" here ?

Reply:
The reason I don't do this is because err won't be set in this scope for some error conditions that cause an early return, eg, if I understand go correctly:
```
	if err := tpm.PolicyCommandCode(policySession, tpm2.CommandNVIncrement); err != nil {
		return nil, nil, xerrors.Errorf("cannot execute assertion to initialize NV index: %w", err)
	}
```

Reply:
Thanks for the example, this will work if you change the signature of the  func createPinNVIndex(...) (a *tpm2.NVPublic, b tpm2.DigestList, err error) so that it has named return attrs. I.e. (..., err error). See https://play.golang.org/p/roZjNb5B5OE.
(but not a blocker or anything it just seems like the extra var could be avoided if we wanted).